File: src/main/java/architecture/goldenboughs_lib/common/payload/toc/GeckolibAnimationSynchroPayload.java
package architecture.goldenboughs_lib.common.payload.toc;

import architecture.goldenboughs_lib.common.payload.api.ToClientPayload;
import architecture.goldenboughs_lib.core.GoldenBoughsLib;
import architecture.goldenboughs_lib.mixin.geckolib.AnimationControllerAccessorMixin;
import io.netty.buffer.ByteBuf;
import net.minecraft.network.codec.ByteBufCodecs;
import net.minecraft.network.codec.StreamCodec;
import net.minecraft.network.protocol.common.custom.CustomPacketPayload;
import net.minecraft.world.entity.player.Player;
import software.bernie.geckolib.animatable.GeoAnimatable;
import software.bernie.geckolib.util.GeckoLibUtil;

public record GeckolibAnimationSynchroPayload(
	String syncableId,
	long instanceId,
	String controllerName,
	String animName
) implements ToClientPayload {
	public static final Type<GeckolibAnimationSynchroPayload> TYPE = new Type<>(GoldenBoughsLib.modRl("curio_animation_synchro"));
	public static final StreamCodec<ByteBuf, GeckolibAnimationSynchroPayload> STREAM_CODEC = StreamCodec.composite(
		ByteBufCodecs.STRING_UTF8, GeckolibAnimationSynchroPayload::syncableId,
		ByteBufCodecs.VAR_LONG, GeckolibAnimationSynchroPayload::instanceId,
		ByteBufCodecs.STRING_UTF8, GeckolibAnimationSynchroPayload::controllerName,
		ByteBufCodecs.STRING_UTF8, GeckolibAnimationSynchroPayload::animName,
		GeckolibAnimationSynchroPayload::new);

	public GeckolibAnimationSynchroPayload(GeoAnimatable animatable, long instanceId, String controllerName, String animName) {
		this(GeckoLibUtil.getSyncedSingletonAnimatableId(animatable), instanceId, controllerName, animName);
	}

	@Override
	public void work(Player player) {
		GeoAnimatable animatable = GeckoLibUtil.getSyncedAnimatable(this.syncableId);

		if (animatable != null) {
			var animatableInstanceCache = animatable.getAnimatableInstanceCache();
			var managerForId = animatableInstanceCache.getManagerForId(this.instanceId);
			var animationControllers = managerForId.getAnimationControllers();
			var controller = animationControllers.get(this.controllerName);
			controller.setAnimation(((AnimationControllerAccessorMixin) controller).goldenboughs_lib$getTriggerableAnimations().get(this.animName));
		}
	}

	@Override
	public Type<? extends CustomPacketPayload> type() {
		return TYPE;
	}
}

File: src/main/kotlin/architecture/goldenboughs_lib/common/payload/toc/GeckolibAnimationSynchroPayload.kt
package architecture.goldenboughs_lib.common.payload.toc

import architecture.goldenboughs_lib.api.payload.ToClientPayload
import architecture.goldenboughs_lib.util.LibUtil
import architecture.goldenboughs_lib.util.getTriggerableAnimations
import io.netty.buffer.ByteBuf
import net.minecraft.client.player.AbstractClientPlayer
import net.minecraft.network.codec.ByteBufCodecs
import net.minecraft.network.codec.StreamCodec
import net.minecraft.network.protocol.common.custom.CustomPacketPayload
import net.neoforged.neoforge.network.handling.IPayloadContext
import software.bernie.geckolib.animatable.GeoAnimatable
import software.bernie.geckolib.animation.AnimatableManager
import software.bernie.geckolib.animation.AnimationController
import software.bernie.geckolib.util.GeckoLibUtil

@JvmRecord
data class GeckolibAnimationSynchroPayload(
	@JvmField val syncableId: String,
	@JvmField val instanceId: Long,
	@JvmField val controllerName: String,
	@JvmField val animName: String
) : ToClientPayload {
	constructor(animatable: GeoAnimatable, instanceId: Long, controllerName: String, animName: String) : this(
		GeckoLibUtil.getSyncedSingletonAnimatableId(animatable),
		instanceId,
		controllerName,
		animName
	)

	override fun work(context: IPayloadContext, player: AbstractClientPlayer) {
		val animatable = GeckoLibUtil.getSyncedAnimatable(this.syncableId)

		if (animatable != null) {
			val animatableInstanceCache = animatable.animatableInstanceCache
			val managerForId: AnimatableManager<GeoAnimatable> = animatableInstanceCache.getManagerForId(this.instanceId)
			val animationControllers = managerForId.getAnimationControllers()
			val controller: AnimationController<GeoAnimatable>? = animationControllers[this.controllerName]
			controller?.setAnimation(
				controller.getTriggerableAnimations()[this.animName]
			)
		}
	}

	override fun type(): CustomPacketPayload.Type<out CustomPacketPayload> {
		return TYPE
	}

	companion object {
		@JvmField
		val TYPE: CustomPacketPayload.Type<GeckolibAnimationSynchroPayload> =
			CustomPacketPayload.Type<GeckolibAnimationSynchroPayload>(LibUtil.modRl("curio_animation_synchro"))

		@JvmField
		val STREAM_CODEC: StreamCodec<ByteBuf, GeckolibAnimationSynchroPayload> =
			StreamCodec.composite(
				ByteBufCodecs.STRING_UTF8, GeckolibAnimationSynchroPayload::syncableId,
				ByteBufCodecs.VAR_LONG, GeckolibAnimationSynchroPayload::instanceId,
				ByteBufCodecs.STRING_UTF8, GeckolibAnimationSynchroPayload::controllerName,
				ByteBufCodecs.STRING_UTF8, GeckolibAnimationSynchroPayload::animName,
				::GeckolibAnimationSynchroPayload
			)
	}
}
